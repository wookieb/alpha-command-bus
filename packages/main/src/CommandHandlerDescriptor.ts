import {Command} from 'alpha-command-bus-core';
import {ShapeCommand} from 'alpha-command-bus-command-factory/src/ShapeCommand';

const matchesObject = require('lodash.matches');


export class CommandHandlerDescriptor<TCommand extends Command = Command> {
    constructor(readonly predicate: CommandHandlerDescriptor.Predicate<TCommand>,
                readonly func: CommandHandlerDescriptor.Func<TCommand>) {
        Object.freeze(this);
    }

    static fromFilter<TCommand extends Command = Command>(filter: CommandHandlerDescriptor.Filter<TCommand>,
                                                          func: CommandHandlerDescriptor.Func<TCommand>) {
        return new CommandHandlerDescriptor<TCommand>(
            CommandHandlerDescriptor.filterToPredicate(filter),
            func
        );
    }

    static filterToPredicate(filter: CommandHandlerDescriptor.Filter<any>): CommandHandlerDescriptor.Predicate<any> {
        if (typeof filter === 'string') {
            const commandName = filter;
            return (c: Command) => c.command === commandName;
        }

        if (ShapeCommand.isType(filter)) {
            return (c: Command) => c.command === filter.COMMAND_NAME;
        }

        if (filter instanceof Function) {
            return filter as CommandHandlerDescriptor.Predicate<any>;
        }

        if (typeof filter === 'object') {
            return matchesObject(filter) as CommandHandlerDescriptor.Predicate<any>;
        }

        throw new Error('Command predicate has to be a function, a string, an object or class or ShapeCommand.Shape')
    }
}

export namespace CommandHandlerDescriptor {
    export type Predicate<TCommand extends Command> = (command: TCommand) => boolean;

    export type Func<TCommand extends Command, TResult = any> = (command: TCommand) => Promise<TResult> | TResult;

    export type Filter<TCommand extends Command> = string | object | Predicate<TCommand> | ShapeCommand.Shape;
}
