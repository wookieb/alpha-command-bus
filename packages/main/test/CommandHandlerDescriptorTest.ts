import {CommandHandlerDescriptor} from "@src/CommandHandlerDescriptor";
import * as sinon from 'sinon';

describe('CommandHandlerDescriptor', () => {
    const COMMAND_NAME = 'commandName';

    describe('filterToPredicate', () => {
        it('string as command name', () => {

            const predicate = CommandHandlerDescriptor.filterToPredicate(COMMAND_NAME);

            expect(predicate({command: COMMAND_NAME}))
                .toBeTruthy();

            expect(predicate({command: 'random'}))
                .toBeFalsy();
        });

        it('object as object match', () => {
            const predicate1 = CommandHandlerDescriptor.filterToPredicate({
                command: COMMAND_NAME,
                arg: 1
            });

            const predicate2 = CommandHandlerDescriptor.filterToPredicate({
                command: COMMAND_NAME
            });

            expect(predicate1({command: COMMAND_NAME, arg: 1}))
                .toBeTruthy();

            expect(predicate1({command: COMMAND_NAME, arg: 2}))
                .toBeFalsy();

            expect(predicate2({command: COMMAND_NAME, arg: 1}))
                .toBeTruthy();

            expect(predicate2({command: COMMAND_NAME, arg: 2}))
                .toBeTruthy();

            expect(predicate2({command: COMMAND_NAME}))
                .toBeTruthy();
        });

        it('predicate as command predicate', () => {
            const filter = sinon.stub().returns(true);

            const predicate = CommandHandlerDescriptor.filterToPredicate(filter);

            const command = {command: COMMAND_NAME};
            expect(predicate(command))
                .toBeTruthy();

            sinon.assert.calledWith(filter, command);
        });

        it('throws error if filter has invalid type', () => {
            expect(() => {
                CommandHandlerDescriptor.filterToPredicate(false as any)
            })
                .toThrowErrorMatchingSnapshot();
        });
    });

    it('creating from filter', () => {
        const predicate = sinon.stub().returns(true);
        const func = sinon.stub();
        const descriptor = CommandHandlerDescriptor.fromFilter(predicate, func);

        expect(descriptor)
            .toEqual(
                new CommandHandlerDescriptor(predicate, func)
            );
    });
});