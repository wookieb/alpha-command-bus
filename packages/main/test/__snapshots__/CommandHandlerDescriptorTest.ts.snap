// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandHandlerDescriptor filterToPredicate throws error if filter has invalid type 1`] = `"Command predicate has to be a function, a string, an object or class or ShapeCommand.Shape"`;
